impeller/toolkit/interop/impeller.cc

```diff
@@ -817,4 +826,225 @@ void ImpellerPaintSetMaskFilter(ImpellerPaint paint,
+void ImpellerParagraphStyleSetFontSize(ImpellerParagraphStyle paragraph_style,
+                                       float size) {
+  GetPeer(paragraph_style)->SetFontSize(size);
+}
+
+IMPELLER_EXTERN_C
+void ImpellerParagraphStyleSetHeight(ImpellerParagraphStyle paragraph_style,
+                                     float height) {
+  GetPeer(paragraph_style)->SetHeight(height);
+}
+
+IMPELLER_EXTERN_C
+void ImpellerParagraphStyleSetTextAlignment(
+    ImpellerParagraphStyle paragraph_style,
+    ImpellerTextAlignment align) {
+  GetPeer(paragraph_style)->SetTextAlignment(ToTxtType(align));
+}
+
+IMPELLER_EXTERN_C
+void ImpellerParagraphStyleSetTextDirection(
+    ImpellerParagraphStyle paragraph_style,
+    ImpellerTextDirection direction) {
+  GetPeer(paragraph_style)->SetTextDirection(ToTxtType(direction));
+}
+
+IMPELLER_EXTERN_C
+void ImpellerParagraphStyleSetMaxLines(ImpellerParagraphStyle paragraph_style,
+                                       uint32_t max_lines) {
+  GetPeer(paragraph_style)->SetMaxLines(max_lines);
+}
+
+IMPELLER_EXTERN_C
+void ImpellerParagraphStyleSetLocale(ImpellerParagraphStyle paragraph_style,
+                                     const char* locale) {
+  GetPeer(paragraph_style)->SetLocale(ReadString(locale));
+}
+
+IMPELLER_EXTERN_C
+void ImpellerDisplayListBuilderDrawParagraph(ImpellerDisplayListBuilder builder,
+                                             ImpellerParagraph paragraph,
+                                             const ImpellerPoint* point) {
+  GetPeer(builder)->DrawParagraph(*GetPeer(paragraph), ToImpellerType(*point));
+}
+
+IMPELLER_EXTERN_C ImpellerParagraphBuilder ImpellerParagraphBuilderNew(
+    ImpellerTypographyContext context) {
+  auto builder = Create<ParagraphBuilder>(*GetPeer(context));
+  if (!builder->IsValid()) {
+    VALIDATION_LOG << "Could not create valid paragraph builder.";
+    return nullptr;
+  }
+  return builder.Leak();
+}
+
+IMPELLER_EXTERN_C
+void ImpellerParagraphBuilderRetain(
+    ImpellerParagraphBuilder paragraph_builder) {
+  ObjectBase::SafeRetain(paragraph_builder);
+}
+
+IMPELLER_EXTERN_C
+void ImpellerParagraphBuilderRelease(
+    ImpellerParagraphBuilder paragraph_builder) {
+  ObjectBase::SafeRelease(paragraph_builder);
+}
+
+IMPELLER_EXTERN_C
+void ImpellerParagraphBuilderPushStyle(
+    ImpellerParagraphBuilder paragraph_builder,
+    ImpellerParagraphStyle style) {
+  GetPeer(paragraph_builder)->PushStyle(*GetPeer(style));
+}
+
+IMPELLER_EXTERN_C
+void ImpellerParagraphBuilderPopStyle(
+    ImpellerParagraphBuilder paragraph_builder) {
+  GetPeer(paragraph_builder)->PopStyle();
+}
+
+IMPELLER_EXTERN_C
+void ImpellerParagraphBuilderAddText(ImpellerParagraphBuilder paragraph_builder,
+                                     const uint8_t* data,
+                                     uint32_t length) {
+  if (data == nullptr) {
+    length = 0;
+  }
+  if (length == 0) {
+    return;
+  }
+  GetPeer(paragraph_builder)
+      ->AddText(fml::Utf8ToUtf16(
```

Comment:
The skia paragraph builder supports utf8 strings too. We only have the utf16 API exposed because that is the format of dart strings

Reply:
Cool. I'll wire that up directly instead of doing this conversion.

Reply:
I thought UTF8 was a better default.

Reply:
Done.